### PR TITLE
Shipping method validation

### PR DESCRIPTION
**Pull request by AmooAti**
## Issue Reference
#5305 

## Description
Checker added to saveShippingMethod.

## How To Test This?
Add a product to cart disable a shipping method and try to use save-shipping API, you will get a 400 error.

### CODE REVIEW

File: packages/Webkul/Checkout/src/Cart.php [hunk 1]
```diff
@@ -446,6 +446,19 @@ public function saveShippingMethod($shippingMethodCode): bool
...
+            foreach ($shippingMethod['rates'] as $rate)
+            if ($rate->method == $shippingMethodCode) {
+                $isMethodAvailable = true;
+                break 2;
+            }
+        }
+        if (!$isMethodAvailable) {
+            return false;
+        }
+
```
Comment (devansh-webkul) [inline packages/Webkul/Checkout/src/Cart.php:461]: This can be more shorten, if you check this method also,
https://github.com/bagisto/bagisto/blob/a832c34644434a5149f1ad9df8e48ade8e9b999b/packages/Webkul/Shipping/src/Shipping.php#L129

**Review by devansh-webkul — CHANGES_REQUESTED** [review]

File: packages/Webkul/Checkout/src/Cart.php (hunk 1 above)
Comment (AmooAti) [reply packages/Webkul/Checkout/src/Cart.php:461]: Hi @devansh-webkul
There's a point I can't understand so for example, we want flatrate to be set on cart, the code that sends to method is flatrate_flatrate (shipping method) but in the code you suggest it uses shipping class string that is saved in the config and creates a class of that so for my example it will be 'flatrate'.
but I don't have flatrate in here so that I can't use something like Config::get("carriers.{$shippingMethodCode}") and then get class of that 
so in summary my question is how to get a shipping class with shipping method code like flatrate_flatrate?
we can use other approaches like adding the shipping method code to the keys of the array where returns in getShippingMethod facade  and just check that if the key exists

Comment (devansh-webkul) [reply packages/Webkul/Checkout/src/Cart.php:461]: Hi @AmooAti,

You can create one more protected property named as `method` which will get access same as `code`.

Then you can update this method i.e. `getShippingMethods` and update the key like,

~~~php
'code' => $this->getCode(),
'method' => $this->getMethod(),
~~~

Now the existing method will be updated, now you can check with `in_array`.

**Comment by AmooAti** [comment]
@devansh-webkul Thanks for your guide,
I added the property to the abstract class and also add the method to methods array keys and values and checked for references on this method and seems there's no error but it will break all third-party packages.
I think there should be an update on documentation example codes and also bagisto package generator.
I can handle those if this will be the final code.

**Comment by devansh-webkul** [comment]
Yes, this will break the third-party package, I need to discuss this first with the team.

File: packages/Webkul/Shipping/src/Carriers/AbstractShipping.php [hunk 2]
```diff
@@ -2,14 +2,15 @@
 
 namespace Webkul\Shipping\Carriers;
 
-use Illuminate\Support\Facades\Config;
-
 abstract class AbstractShipping
 {
+    protected $code;
```
Comment (devansh-webkul) [inline packages/Webkul/Shipping/src/Carriers/AbstractShipping.php:7]: PHPDoc Missing.

File: packages/Webkul/Shipping/src/Carriers/FlatRate.php [hunk 3]
```diff
@@ -14,11 +14,13 @@
...
     /**
-     * Payment method code
+     * Shipping method code
      *
      * @var string
      */
-    protected $code  = 'flatrate';
+    protected $code = 'flatrate';
+
+    protected $method = 'flatrate_flatrate';
```
Comment (devansh-webkul) [inline packages/Webkul/Shipping/src/Carriers/FlatRate.php:23]: Same here.

File: packages/Webkul/Shipping/src/Carriers/Free.php [hunk 4]
```diff
@@ -13,11 +13,13 @@
...
     /**
-     * Payment method code
+     * Shipping method code
      *
      * @var string
      */
-    protected $code  = 'free';
+    protected $code = 'free';
+
+    protected $method = 'free_free';
```
Comment (devansh-webkul) [inline packages/Webkul/Shipping/src/Carriers/Free.php:22]: Same here.

File: packages/Webkul/Checkout/src/Cart.php [hunk 5]
```diff
@@ -446,6 +446,10 @@ public function saveShippingMethod($shippingMethodCode): bool
             return false;
         }
 
+        if (!array_key_exists($shippingMethodCode, Shipping::getShippingMethods())) {
```
Comment (devansh-webkul) [inline packages/Webkul/Checkout/src/Cart.php:449]: One space after `!`.

**Review by devansh-webkul — CHANGES_REQUESTED** [review]

**Comment by AmooAti** [comment]
Hi, @devansh-webkul 
Thank you, I fixed the code style and doctypes.
thanks for backward compatibility but there's a point if third-parties follow the document instructions we can automatically generate $method with `$code_$code` and put an extra version compare on top of checker in saveShipping method in cart

**Comment by devansh-webkul** [comment]
Yes. this can be done.

**Comment by AmooAti** [comment]
So shall i go with backward compatibility  or you handle that?

**Comment by devansh-webkul** [comment]
No problem, i will manage.

**Comment by AmooAti** [comment]
Ok, thanks 💜
